src/Microsoft.AspNetCore.Mvc.Core/Internal/FilterActionInvoker.cs

```diff
@@ -262,24 +262,24 @@ private Task InvokeAllResourceFiltersAsync()
 
                     await item.FilterAsync.OnResourceExecutionAsync(_resourceExecutingContext, InvokeResourceFilterAwaitedAsync);
 
-                    _diagnosticSource.AfterOnResourceExecution(_resourceExecutedContext, item.FilterAsync);
```

Comment:
I actually found another bug of the same variety by adding these asserts. If an async filter short-circuits, then `_resourceExecutedContext` hasn't been instantiated yet.